README.md
catbox-memcached
================

Memcached adapter for catbox

[![Build Status](https://secure.travis-ci.org/hapijs/catbox-memcached.png)](http://travis-ci.org/hapijs/catbox-memcached)

Lead Maintainer: [Tony Di Benedetto](https://github.com/iniva)

### Options

- `host` - the Memcache server hostname. Defaults to `127.0.0.1`. **Cannot be used with `location`.**
- `port` - the Memcache server port. Defaults to `11211`. **Cannot be used with `location`.**
- `location` - the Memcache server hostname and port. Defaults to `127.0.0.1:11211`. Can be a String, Array, or an Object as per [node-memcached location specification](https://github.com/3rd-Eden/node-memcached#server-locations).


CHANGELOG.md
Breaking changes are documented using GitHub issues, see [issues labeled "release notes"](https://github.com/hapijs/catbox-memcached/issues?q=is%3Aissue+label%3A%22release+notes%22).

If you want changes of a specific minor or patch release, you can browse the [GitHub milestones](https://github.com/hapijs/catbox-memcached/milestones?state=closed&direction=asc&sort=due_date).


test/index.js
'use strict';

const Catbox = require('catbox');
const Code = require('code');
const Lab = require('lab');
const CatboxMemcached = require('../');


const { describe, it } = exports.lab = Lab.script();
const { expect } = Code;


describe('Client', () => {

    it('creates a new connection', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        expect(client.isReady()).to.equal(true);
    });

    it('closes the connection', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        expect(client.isReady()).to.equal(true);

        client.stop();
        expect(client.isReady()).to.equal(false);
    });

    it('gets an item after setting it', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        const key = { id: 'x', segment: 'test' };
        await client.set(key, '123', 500);

        const result = await client.get(key);
        expect(result.item).to.equal('123');
    });

    it('fails setting an item circular references', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        const key = { id: 'x', segment: 'test' };
        const value = { a: 1 };
        value.b = value;

        try {
            await client.set(key, value, 10);
        }
        catch (err) {
            expect(err.message).to.exist();
            expect(err.message).to.equal('Converting circular structure to JSON');
        }
    });

    it('fails setting an item with very long ttl', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        const key = { id: 'x', segment: 'test' };

        try {
            await client.set(key, '123', Math.pow(2, 31));
        }
        catch (err) {
            expect(err).to.exist();
            expect(err.message).to.equal('Invalid ttl (greater than 2147483647)');
        }
    });

    it('ignored starting a connection twice on same event', async () => {

        const client = new Catbox.Client(CatboxMemcached);

        const start = async () => {

            await client.start();
            expect(client.isReady()).to.equal(true);
        };

        await start();
        await start();
    });

    it('ignored starting a connection twice chained', async () => {

        const client = new Catbox.Client(CatboxMemcached);

        await client.start();
        expect(client.isReady()).to.equal(true);

        await client.start();
        expect(client.isReady()).to.equal(true);
    });

    it('returns not found on get when using null key', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        const result = await client.get(null);
        expect(result).to.equal(null);
    });

    it('returns not found on get when item expired', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        const key = { id: 'x', segment: 'test' };

        await client.set(key, 'x', 1);
        await new Promise((resolve) => {

            setTimeout(async () => {

                const result = await client.get(key);
                expect(result).to.equal(null);
                resolve();
            }, 2);
        });
    });

    it('returns error on set when using null key', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        await expect(client.set(null, {}, 1000)).to.reject();
    });

    it('returns error on get when using invalid key', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        await expect(client.get({})).to.reject();
    });

    it('returns error on drop when using invalid key', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        await expect(client.drop({})).to.reject();
    });

    it('returns error on set when using invalid key', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        await expect(client.set({}, {}, 1000)).to.reject();
    });

    it('ignores set when using non-positive ttl value', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        const key = { id: 'x', segment: 'test' };
        await client.set(key, 'y', 0);
    });

    it('returns error on drop when using null key', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        await client.start();

        await expect(client.drop(null)).to.reject();
    });

    it('returns error on get when stopped', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        client.stop();

        const key = { id: 'x', segment: 'test' };
        await expect(client.get(key)).to.reject();
    });

    it('returns error on set when stopped', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        client.stop();

        const key = { id: 'x', segment: 'test' };
        await expect(client.set(key, 'y', 1)).to.reject();
    });

    it('returns error on drop when stopped', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        client.stop();

        const key = { id: 'x', segment: 'test' };
        await expect(client.drop(key)).to.reject();
    });

    it('returns error on missing segment name', () => {

        const config = {
            expiresIn: 50000
        };

        const fn = () => {

            const client = new Catbox.Client(CatboxMemcached);
            new Catbox.Policy(config, client, '');
        };

        expect(fn).to.throw(Error);
    });

    it('returns error on bad segment name', () => {

        const config = {
            expiresIn: 50000
        };

        const fn = () => {

            const client = new Catbox.Client(CatboxMemcached);
            new Catbox.Policy(config, client, 'a\0b');
        };

        expect(fn).to.throw(Error);
    });

    it('returns error when cache item dropped while stopped', async () => {

        const client = new Catbox.Client(CatboxMemcached);
        client.stop();

        await expect(client.drop('a')).to.reject();
    });

    it('throws an error if not created with new', () => {

        const fn = () => {

            CatboxMemcached();
        };

        expect(fn).to.throw(Error);
    });

    describe('constructor()', () => {

        it('takes location as a string', () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            expect(memcache.settings.location).to.equal(options.location);
        });

        it('takes location as an array', () => {

            const options = {
                location: ['127.0.0.1:11211']
            };
            const memcache = new CatboxMemcached(options);

            expect(memcache.settings.location).to.equal(options.location);
        });

        it('takes location as an object', () => {

            const options = {
                location: {
                    '127.0.0.1:11211': 1
                }
            };
            const memcache = new CatboxMemcached(options);

            expect(memcache.settings.location).to.equal(options.location);
        });

        it('supports using defaults if no options are passed in', () => {

            let memcache = null;
            const fn = () => {

                memcache = new CatboxMemcached();
            };

            expect(fn).to.not.throw();
            expect(memcache.settings.location).to.equal('127.0.0.1:11211');
        });

        it('throws an error if given location and host/port', () => {

            const fn = () => {

                new CatboxMemcached({
                    location: '127.0.0.1:11211',
                    host: '127.0.0.1',
                    port: 11211
                });
            };

            expect(fn).to.throw(Error);
        });

        it('throws an error if given location and host', () => {

            const fn = () => {

                new CatboxMemcached({
                    location: '127.0.0.1:11211',
                    host: '127.0.0.1'
                });
            };

            expect(fn).to.throw(Error);
        });

        it('throws an error if given location and port', () => {

            const fn = () => {

                new CatboxMemcached({
                    location: '127.0.0.1:11211',
                    port: 11211
                });
            };

            expect(fn).to.throw(Error);
        });

    });

    describe('start()', () => {

        it('sets client when the connection succeeds', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            await memcache.start();
            expect(memcache._client).to.exist();
        });

        it('sets isReady to true when the connection succeeds', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            await memcache.start();
            expect(memcache.isReady()).to.be.true();
        });

        it('reuses the client when a connection is already started', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };

            const memcache = new CatboxMemcached(options);

            await memcache.start();

            const client = memcache.client;

            await memcache.start();
            expect(client).to.equal(memcache.client);
        });

    });

    describe('validateSegmentName()', () => {

        it('returns an error when the name is empty', () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            expect(() => memcache.validateSegmentName('')).to.throw('Empty string');
        });

        it('returns an error when the name has a null character', () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            expect(() => memcache.validateSegmentName('\0test')).to.throw();
        });

        it('returns an error when the name has a space character', () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            expect(() => memcache.validateSegmentName(' test')).to.throw();
        });

        it('returns an error when the name has a tab character', () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            expect(() => memcache.validateSegmentName('\ttest')).to.throw();
        });

        it('returns an error when the name has a new line character', () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            expect(() => memcache.validateSegmentName('\ntest')).to.throw();
        });

        it('returns an error when the name is too long', () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            let tooLongName = '';

            for (; tooLongName.length < 300;) {
                tooLongName += Math.random().toString(36).substring(2, 15);
            }

            expect(() => memcache.validateSegmentName(tooLongName)).to.throw();
        });

        it('returns null when the name is valid', () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            expect(memcache.validateSegmentName('valid')).to.equal(null);
        });

    });

    describe('get()', () => {

        it('returns a rejected promise when the connection is closed', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            await expect(memcache.get('test')).to.reject('Connection is not ready');
        });

        it('returns a null item when it doesn\'t exist', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            await memcache.start();
            const result = await memcache.get('');

            expect(result).to.equal(null);
        });

        it('returns a rejected promise when there is an error returned from client', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            memcache.start();

            memcache._client.get = (item, callback) => callback(new Error('Boom'));

            await expect(memcache.get(key)).to.reject('Boom');
        });

        it('returns a rejected promise when there is an error returned from parsing the result', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            memcache.start();

            memcache._client.get = (item, callback) => callback(null, '{"invalid": "json"');

            await expect(memcache.get(key)).to.reject('Bad envelope content');
        });

        it('returns a rejected promise when there is an error with the envelope structure', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            memcache.start();

            memcache._client.get = (item, callback) => callback(null, '{"wrong": "structure"}');

            await expect(memcache.get(key)).to.reject('Incorrect envelope structure');
        });

        it('is able to retrieve an object that\'s stored when connection is started', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            await memcache.start();
            await memcache.set(key, 'myValue', 200);
            const result = await memcache.get(key);

            expect(result.item).to.equal('myValue');
        });

    });

    describe('set()', () => {

        it('returns a rejected promise when the connection is closed', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            await expect(memcache.set(key, 'test1', 3600)).to.reject('Connection is not ready');
        });

        it('returns a rejected promise when there is an error returned from setting an item', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            memcache.start();

            memcache._client.set = (targetKey, value, ttl, callback) => callback(new Error('Error'));

            await expect(memcache.set(key, 'value', 0)).to.reject('Error');
        });

        it('doesn\'t return an error when the set succeeds', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            await memcache.start();
            const result = await memcache.set(key, 'test1', 3600);

            expect(result).to.not.exist();
        });

    });

    describe('drop()', () => {

        it('returns a rejected promise when the connection is closed', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            await expect(memcache.drop(key)).to.reject('Connection is not ready');
        });

        it('doesn\'t return an error when the drop succeeds', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            await memcache.start();
            const result = await memcache.drop(key);

            expect(result).to.not.exist();
        });

        it('returns a rejected promise when there is an error returned from dropping an item', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);
            const key = {
                id: 'test',
                segment: 'test'
            };

            await memcache.start();

            memcache._client.del = (targetKey, callback) => callback(new Error('Error'));

            await expect(memcache.drop(key)).to.reject('Error');
        });

    });

    describe('stop()', () => {

        it('sets the client to null', async () => {

            const options = {
                location: '127.0.0.1:11211'
            };
            const memcache = new CatboxMemcached(options);

            await memcache.start();
            memcache.stop();

            expect(memcache._client).to.be.equal(null);
        });

    });

});


examples/hapi.js
'use strict';

// After starting this example load http://localhost:8080
// hit refresh, you will notice that it loads the response
// from cache for the first 5 seconds and then reloads the cache

// Load modules
const CatboxMemcached = require('..');
const Hapi = require('hapi');

// create server instance with Memcached cache
const server = Hapi.server({
    host: '0.0.0.0',
    port: 8080,
    cache: [
        {
            name: 'memcached',
            engine: CatboxMemcached,
            host: '127.0.0.1',
            partition: 'example-cache'
        }
    ]
});

// define a cache segment to store cache items
const Cache = server.cache({ segment: 'examples', expiresIn: 1000 * 5 });

// wildcard route that responds all requests
// either with data from cache or default string
server.route({
    method: 'GET',
    path: '/{path*}',
    handler: async (request, h) => {

        const key = {
            segment: 'examples',
            id: 'myExample'
        };

        // get item from cache segment
        const cached = await Cache.get(key);

        if (cached) {
            return `From cache: ${cached.item}`;
        }

        // fill cache with item
        await Cache.set(key, { item: 'my example' }, 5000);

        return 'my example';
    }
});

(async () => {

    try {
        await server.start();
    }
    catch (err) {
        console.log(err);
        process.exit(1);
    }

    console.log(`Server running at: ${server.info.uri}`);
})();


examples/memcached.js
'use strict';

// After starting this example load http://localhost:8080 and hit refresh.
// You will notice that it loads the response from cache
// for the first 5 seconds and then reloads the cache

// Load modules
const Catbox = require('catbox');
const Http = require('http');


// Declare internals
const internals = {};

internals.handler = async (req, res) => {

    try {
        const item = await internals.getResponse();
        res.writeHead(200, { 'Content-Type': 'text/plain' });
        res.end(item);
    }
    catch (ignoreErr) {
        res.writeHead(500);
        res.end();
    }
};


internals.getResponse = async () => {

    const key = {
        segment: 'example',
        id: 'myExample'
    };

    const cacheValue = 'my example';
    const ttl = 10000; // How long item will be cached in milliseconds

    const cached = await internals.client.get(key);

    if (cached) {
        return `From cache: ${cached.item}`;
    }

    await internals.client.set(key, cacheValue, ttl);

    return cacheValue;
};

internals.startCache = async () => {

    const options = {
        partition: 'examples',
        location: '127.0.0.1:11211'
    };
    try {
        internals.client = new Catbox.Client(require('../'), options);  // Replace require('../') with 'catbox-memcached' in your own code
        await internals.client.start();
    }
    catch (error) {
        console.log('error');
        console.log(error);
        process.exit();
    }
};

internals.startServer = () => {

    const server = Http.createServer(internals.handler);
    server.listen(8080);
};

(async () => {

    await internals.startCache();
    internals.startServer();
    console.log('Server started at http://localhost:8080/');
})();

// In case you face unhandled rejection errors
process.on('unhandledRejection', (error) => {

    console.log(error);
    process.exit();
});


lib/index.js
'use strict';

const Hoek = require('hoek');
const Boom = require('boom');
const Memcache = require('memcached');


const internals = {
    defaults: {
        host: '127.0.0.1',
        port: 11211,
        timeout: 1000,
        idle: 1000
    }
};


exports = module.exports = internals.Connection = class {

    constructor(options = {}) {

        Hoek.assert(!(options.location && (options.host || options.port)), 'Cannot specify both location and host/port when using memcached');

        this.settings = Hoek.applyToDefaults(internals.defaults, options);

        if (!this.settings.location) {
            this.settings.location = `${this.settings.host}:${this.settings.port}`;
        }

        delete this.settings.port;
        delete this.settings.host;

        this._client = null;
        this.isConnected = false;
    }

    start() {

        if (this._client) {
            return;
        }

        this._client = new Memcache(this.settings.location, this.settings);
        this.isConnected = true;
    }

    stop() {

        if (this._client) {
            this._client.end();
            this._client = null;
            this.isConnected = false;
        }
    }

    isReady() {

        return this.isConnected;
    }

    validateSegmentName(name) {

        if (!name) {
            throw new Boom('Empty string');
        }

        // https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L44-L49

        if (name.indexOf('\0') !== -1) {
            throw new Boom('Includes null character');
        }

        if (name.match(/\s/g)) {
            throw new Boom('Includes spacing character(s)');
        }

        const { partition = '' } = this.settings;

        if (name.length + partition.length > 250) {
            throw new Boom('Segment and partition name lengths exceeds 250 characters');
        }

        return null;
    }

    get(key) {

        if (!this.isConnected) {
            return Promise.reject(new Boom('Connection is not ready'));
        }

        return new Promise((resolve, reject) => {

            this._client.get(this.generateKey(key), (err, result) => {

                if (err) {
                    return reject(err);
                }

                if (!result) {
                    return resolve(null);
                }

                try {
                    var envelope = JSON.parse(result);
                }
                catch (err) {
                    return reject(new Boom('Bad envelope content'));
                }

                if (!envelope.item ||
                    !envelope.stored) {

                    return reject(new Boom('Incorrect envelope structure'));
                }

                resolve(envelope);
            });
        });
    }

    set(key, value, ttl) {

        if (!this.isConnected) {
            return Promise.reject(new Boom('Connection is not ready'));
        }

        const envelope = {
            item: value,
            stored: Date.now(),
            ttl
        };

        const cacheKey = this.generateKey(key);

        try {
            var stringifiedEnvelope = JSON.stringify(envelope);
        }
        catch (err) {
            return Promise.reject(new Boom(err.message));
        }

        const ttlSec = Math.max(1, Math.floor(ttl / 1000));

        return new Promise((resolve, reject) => {

            this._client.set(cacheKey, stringifiedEnvelope, ttlSec, (err) => {

                if (err) {
                    return reject(new Boom(err.message));
                }

                resolve();
            });
        });
    }

    drop(key) {

        if (!this.isConnected) {
            return Promise.reject(new Boom('Connection is not ready'));
        }

        return new Promise((resolve, reject) => {

            this._client.del(this.generateKey(key), (err) => {

                if (err) {
                    return reject(new Boom(err.message));
                }

                resolve();
            });
        });
    };

    generateKey(key) {

        const { partition = '' } = this.settings;
        const { segment, id } = key;
        let generatedKey = `${encodeURIComponent(segment)}:${encodeURIComponent(id)}`;

        if (partition !== '') {
            generatedKey = `${encodeURIComponent(partition)}:${generatedKey}`;
        }

        return generatedKey;
    };
};
